src/linear_algebra/free_module.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import linear_algebra.basis
+import logic.small
+import linear_algebra.direct_sum.finsupp
+
+/-!
+
+# Free modules
+
+We introduce a class `free_module R M`, for `R` a `semiring` and `M` an `R`-module and we provide
+several basic instances for this class.
+
+## Main definition
+
+* `module.free R M` : the class of free `R`-modules.
+
+-/
+
+universes u v w z
+
+variables (R : Type u) (M : Type v) (N : Type z)
+
+open_locale tensor_product
+
+section basic
+
+variables [semiring R] [add_comm_monoid M] [module R M]
+
+/-- `finite_free R M` is the statement that the `R`-module `R` is free of finite rank.-/
+class module.free : Prop :=
+(exists_basis [] : nonempty (Σ (I : Type v), basis I R M))
+
+lemma module.free_def [small.{w} M] : module.free R M ↔ ∃ (I : Type w), nonempty (basis I R M) :=
+⟨ λ h, ⟨shrink (set.range h.exists_basis.some.2),
+    ⟨(basis.reindex_range h.exists_basis.some.2).reindex (equiv_shrink _)⟩⟩,
+  λ h, ⟨(nonempty_sigma.2 h).map $ λ ⟨i, b⟩, ⟨set.range b, b.reindex_range⟩⟩⟩
+
+lemma module.free_iff_set : module.free R M ↔ ∃ (S : set M), nonempty (basis S R M) :=
+⟨λ h, ⟨set.range h.exists_basis.some.2, ⟨basis.reindex_range h.exists_basis.some.2⟩⟩,
+    λ ⟨S, hS⟩, ⟨nonempty_sigma.2 ⟨S, hS⟩⟩⟩
+
+variables {R M}
+
+lemma module.free.of_basis {ι : Type w} (b : basis ι R M) : module.free R M :=
+(module.free_def R M).2 ⟨set.range b, ⟨b.reindex_range⟩⟩
+
+end basic
+
+namespace module.free
+
+section semiring
+
+variables (R M) [semiring R] [add_comm_monoid M] [module R M] [module.free R M]
+variables [add_comm_monoid N] [module R N]
+
+/-- If `[finite_free R M]` then `choose_basis_index R M` is the `ι` which indexes the basis
+  `ι → M`. -/
+@[nolint has_inhabited_instance] def choose_basis_index := (exists_basis R M).some.1
+
+/-- If `[finite_free R M]` then `choose_basis : ι → M` is the basis.
+Here `ι = choose_basis_index R M`. -/
+noncomputable def choose_basis : basis (choose_basis_index R M) R M := (exists_basis R M).some.2
+
+/-- The isomorphism `M ≃ₗ[R] (choose_basis_index R M →₀ R)`. -/
+noncomputable def repr : M ≃ₗ[R] (choose_basis_index R M →₀ R) := (choose_basis R M).repr
+
+@[priority 100]
+instance no_zero_smul_divisors [no_zero_divisors R] : no_zero_smul_divisors R M :=
+basis.no_zero_smul_divisors $ choose_basis R M
```

Comment:
I think you can do this one without choice, since you're constructing a prop

Reply:
Can you elaborate a little more on this? I am not sure what I should do...

Reply:
```suggestion
let ⟨⟨_, b⟩⟩ := exists_basis R M in b.no_zero_smul_divisors
```

Reply:
I mean ultimately the instance uses `choice` whether you do this or not, so up to you if you think this is actually better.